### PR TITLE
(#98) Allow overriding devmode

### PR DESCRIPTION
**Pull request by sotiredofyours**
Provide the ability to create instance of Paths in the dev mode.

### CODE REVIEW

**Comment by sotiredofyours** [comment]
@ForNeVeR Take a look at the way I have modified Paths, please. I want to make sure I understand your idea (2nd option) correctly. 

And, is it correct to keep a new instance in each class that needs Paths like I did in ArchiveManager? Maybe there is some more convenient way. Advise me.

File: Source/Shared/Paths.cs [hunk 1]
```diff
@@ -3,36 +3,43 @@
...
     private const string DevModeMarkerFileName = ".bloodmasters.dev.marker";
     private const string DevSolutionRootMarkerFileName = ".bloodmasters.solution-root.marker";
     private const string TargetFrameworkForExecutables = "net7.0-windows";
-    private const string BuildConfiguration =
+    private readonly string _buildConfiguration =
 #if DEBUG
         "Debug";
 #else
         "Release";
 #endif
```
Comment (ForNeVeR) [inline Source/Shared/Paths.cs:16]: Nope, let's leave this as a `const`.

File: Source/Shared/Paths.cs [hunk 2]
```diff
@@ -3,36 +3,43 @@
...
 
+    public Paths(bool isDev = false)
+    {
+        if (isDev)
+            _buildConfiguration = "Debug";
+    }
```
Comment (ForNeVeR) [inline Source/Shared/Paths.cs:22]: This flag is supposed to override the `IsDevModeBuild`, not the `_buildConfiguration`.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

**Comment by ForNeVeR** [comment]
> And, is it correct to keep a new instance in each class that needs Paths like I did in ArchiveManager?

Let's create a default `Paths.Instance` and use it everywhere.

The default instance should detect the dev mode based on the marker file in the file system (as it already does), while a customly-created instance should use whatever `isDevMode` value is provided in the constructor.

**Comment by sotiredofyours** [comment]
@ForNeVeR I've redone the logic within Paths. Now everything works through Paths.Instance. 
 
I also separated the tests that checked `ConfigDirPath`.  At the moment the `devMode` test fails. I'm trying to figure out the reason. Maybe I missed something and it shouldn't work like that?

File: Source/Shared/Paths.cs [hunk 3]
```diff
@@ -3,36 +3,43 @@
...
+    public static Paths Instance { get; } = new ();
+
     /// <summary>Directory of the entry binary.</summary>
     /// <remarks>In dev mode, it looks like <c>Source/Client/bin/Debug/net7.0-windows</c>.</remarks>
-    private static readonly string AppBaseDir = GetAppBaseDir();
-    private static string GetAppBaseDir()
+    private string AppBaseDir => GetAppBaseDir();
+    private string GetAppBaseDir()
     {
         var modulePath = Assembly.GetEntryAssembly()?.Location ?? Process.GetCurrentProcess().MainModule?.FileName;
```
Comment (sotiredofyours) [inline Source/Shared/Paths.cs:30]: I see what this is about. Here we get to the TestRunner package, not our application. Should I just delete this test and revert to the previous version?

File: Source/Shared/Paths.cs [hunk 4]
```diff
@@ -3,36 +3,43 @@
...
 namespace CodeImp.Bloodmasters;
 
-public static class Paths
+public class Paths
 {
     private const string DevModeMarkerFileName = ".bloodmasters.dev.marker";
     private const string DevSolutionRootMarkerFileName = ".bloodmasters.solution-root.marker";
     private const string TargetFrameworkForExecutables = "net7.0-windows";
-    private const string BuildConfiguration =
+    private const string _buildConfiguration =
```
Comment (ForNeVeR) [inline Source/Shared/Paths.cs:11]: Please roll back this rename.

File: Source/Shared/Paths.cs (hunk 3 above)
Comment (ForNeVeR) [reply Source/Shared/Paths.cs:30]: > Here we get to the TestRunner package, not our application. 

This is expected.

File: Source/Shared/Paths.cs [hunk 5]
```diff
@@ -3,36 +3,43 @@
...
+    public Paths(bool isDev = false)
+    {
+        IsDevModeBuild = File.Exists(Path.Combine(AppBaseDir, DevModeMarkerFileName)) || isDev;
+    }
+
+    public static Paths Instance { get; } = new ();
+
     /// <summary>Directory of the entry binary.</summary>
     /// <remarks>In dev mode, it looks like <c>Source/Client/bin/Debug/net7.0-windows</c>.</remarks>
-    private static readonly string AppBaseDir = GetAppBaseDir();
```
Comment (ForNeVeR) [inline Source/Shared/Paths.cs:20]: I propose to make all these members (non-static) readonly fields, and not properties that get re-calculated on every access.

File: Source/Tests/Paths/ConfigDirPathTests.cs [hunk 6]
```diff
@@ -2,25 +2,34 @@ namespace Bloodmasters.Tests.Paths;
 
 public class ConfigDirPathTests
 {
-    [Fact(DisplayName = "ConfigDirPath should be named 'Debug' in dev mode and 'Config' otherwise")]
+    [Fact(DisplayName = "ConfigDirPath should be named 'Config' in release mode")]
```
Comment (ForNeVeR) [inline Source/Tests/Paths/ConfigDirPathTests.cs:5]: Let's call it "production mode", to not confuse with build configurations.

File: Source/Tests/Paths/ConfigDirPathTests.cs [hunk 7]
```diff
@@ -2,25 +2,34 @@ namespace Bloodmasters.Tests.Paths;
...
-        var dirName = Path.GetFileName(CodeImp.Bloodmasters.Paths.ConfigDirPath);
+        var paths = new CodeImp.Bloodmasters.Paths(isDev: true);
+        var dirName = Path.GetFileName(paths.ConfigDirPath);
 
         // Assert
-        Assert.Equal(
-            CodeImp.Bloodmasters.Paths.IsDevModeBuild ? "Debug" : "Config",
-            dirName);
+        Assert.Equal("Debug", dirName);
     }
```
Comment (ForNeVeR) [inline Source/Tests/Paths/ConfigDirPathTests.cs:24]: This whole test is incorrect. `isDev` flag should not control whether `Debug` or `Release` is chosen.

The "dev mode" is when you build the game from sources, either in Debug or Release build configuration.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

**Pull request closed without merging**